consul-lambda-registrator/extension/extension.go

```diff
@@ -0,0 +1,310 @@
+package extension
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/proxy"
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/structs"
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/trace"
+)
+
+type Config struct {
+	ServiceName         string
+	ServiceNamespace    string        `envconfig:"CONSUL_SERVICE_NAMESPACE"`
+	ServicePartition    string        `envconfig:"CONSUL_SERVICE_PARTITION"`
+	ServiceUpstreams    []string      `envconfig:"CONSUL_SERVICE_UPSTREAMS"`
+	MeshGatewayURI      string        `envconfig:"CONSUL_MESH_GATEWAY_URI" required:"true"`
+	ExtensionDataPrefix string        `envconfig:"CONSUL_EXTENSION_DATA_PREFIX" required:"true"`
+	RefreshFrequency    time.Duration `envconfig:"CONSUL_REFRESH_FREQUENCY" default:"5m"`
+
+	Store  ParamGetter
+	Events EventProcessor
+	Logger hclog.Logger
+}
+
+type ParamGetter interface {
+	// Get the value for the given key.
+	Get(ctx context.Context, key string) (string, error)
+}
+
+type EventProcessor interface {
+	// Register the event processor.
+	Register(ctx context.Context, i interface{}) error
+	// ProcessEvents handles events until the provided context is cancelled or an error occurs.
+	ProcessEvents(ctx context.Context) error
+}
+
+type Extension struct {
+	*Config
+	service    structs.Service
+	proxy      *proxy.Server
+	proxyMutex sync.Mutex
+	data       structs.ExtensionData
+	upstreams  []structs.Service
+}
+
+// NewExtension returns an instance of the Extension from the given configuration.
+func NewExtension(cfg *Config) *Extension {
+	ext := &Extension{
+		Config: cfg,
+		service: structs.Service{
+			Name:      cfg.ServiceName,
+			Namespace: cfg.ServiceNamespace,
+			Partition: cfg.ServicePartition,
+		},
+	}
+	return ext
+}
+
+// Serve executes the main processing loop for the extension.
+// It initializes and starts the proxy server and starts monitoring for incoming
+// events from the Lambda runtime.
+// It periodically retrieves the extension data from the parameter store and updates
+// the proxy configuration for the configured upstreams as necessary.
+func (ext *Extension) Serve(ctx context.Context) error {
```

Comment:
Maybe `Start` is a better name for this since it both runs the proxy and handles the events?